File: Mage/StyledPolyline.m
//
//  StyledPolyline.m
//  MAGE
//
//

@import HexColors;

#import "StyledPolyline.h"

@implementation StyledPolyline

+ (StyledPolyline *) generatePolyline:(NSArray *) path {
    NSUInteger numberOfSteps = path.count;
    
    CLLocationCoordinate2D coordinates[numberOfSteps];
    
    for (NSInteger index = 0; index < numberOfSteps; index++) {
        NSNumber *y = path[index][0];
        NSNumber *x = path[index][1];
        coordinates[index] = CLLocationCoordinate2DMake([x doubleValue], [y doubleValue]);
    }
    
    return [StyledPolyline polylineWithCoordinates:coordinates count:path.count];
}

+ (StyledPolyline *) createWithPolyline: (MKPolyline *) polyline {
    StyledPolyline * styledPolyline = [StyledPolyline polylineWithPoints:[polyline points] count:polyline.pointCount];
    [styledPolyline setTitle:polyline.title];
    [styledPolyline setSubtitle:polyline.subtitle];
    [styledPolyline applyDefaultStyle];
    
    return styledPolyline;
}

+ (StyledPolyline *) polylineWithPoints:(const MKMapPoint *)points count:(NSUInteger)count {
    StyledPolyline *styledPolyline = [super polylineWithPoints:points count:count];
    [styledPolyline applyDefaultStyle];

    return styledPolyline;
}

+ (StyledPolyline *) polylineWithCoordinates:(const CLLocationCoordinate2D *)coords count:(NSUInteger)count {
    StyledPolyline *styledPolyline = [super polylineWithCoordinates:coords count:count];
    [styledPolyline applyDefaultStyle];
    
    return styledPolyline;
}

- (void) applyDefaultStyle {
    [self setLineColor:UIColor.blackColor];
    [self setLineWidth:1.0];
}

- (void) lineColorWithHexString: (NSString *) hex andAlpha: (CGFloat) alpha {
    if (hex) {
        _lineColor = [UIColor hx_colorWithHexRGBAString:hex alpha:alpha];
    }
}

- (void) lineColorWithHexString: (NSString *) hex {
    if (hex) {
        _lineColor = [UIColor hx_colorWithHexRGBAString:hex];
    }
}

@end

File: Mage/StyledPolyline.swift
//
//  StyledPolyline.m
//  MAGE
//
//

import CoreLocation

@objc class StyledPolyline : MKPolyline, OverlayRenderable {
    var renderer: MKOverlayRenderer {
        get {
            let renderer = MKPolylineRenderer(polyline: self)
            renderer.strokeColor = lineColor
            renderer.lineWidth = lineWidth
            return renderer
        }
    }
    
    @objc public var lineColor: UIColor = .black
    @objc public var lineWidth: CGFloat = 1.0
    @objc var observationRemoteId: String?
    public var _observation: Observation?
    
    public var observation: Observation? {
        get {
            guard let observationRemoteId = observationRemoteId else {
                return _observation
            }
            
            return Observation.mr_findFirst(byAttribute: "remoteId", withValue: observationRemoteId, in: NSManagedObjectContext.mr_default())
        }
        set {
            if let remoteId = newValue?.remoteId {
                observationRemoteId = remoteId
            } else {
                _observation = newValue
            }
        }
    }
    
    @objc static func generate(path: [[NSNumber]])-> StyledPolyline {
        var coordinates: [CLLocationCoordinate2D] = []
        for point in path {
            coordinates.append(CLLocationCoordinate2D(latitude: point[1].doubleValue, longitude: point[0].doubleValue))
        }
        
        return StyledPolyline(coordinates: coordinates, count: path.count)
    }
    
    @objc static func create(polyline: MKPolyline) -> StyledPolyline {
        let styledPolyline = StyledPolyline(points: polyline.points(), count: polyline.pointCount)
        styledPolyline.title = polyline.title
        styledPolyline.subtitle = polyline.subtitle
        return styledPolyline
    }
    
    @objc public func setLineColor(hex: String, alpha: CGFloat = 1.0) {
        self.lineColor = UIColor(hex: hex)?.withAlphaComponent(alpha) ?? self.lineColor
    }

}
